File: frontend/src/helpers/encoding.js
export function base64_url_encode(input) {
  let output = btoa(input)
  output = output.replaceAll('+', '.')
  output = output.replaceAll('/', '_')
  output = output.replaceAll('=', '-')
  return output
}

export function base64_url_decode(input) {
  let output = input.replaceAll('-', '=')
  output = output.replaceAll('_', '/')
  output = output.replaceAll('.', '+')
  return atob(output)
}

File: internal/domain/encoding.go
package domain

import (
	"encoding/base64"
	"strings"
)

// Base64UrlDecode decodes a base64 url encoded string.
// In comparison to the standard base64 encoding, the url encoding uses - instead of + and _ instead of /
// as well as . instead of =.
func Base64UrlDecode(in string) string {
	in = strings.ReplaceAll(in, "-", "=")
	in = strings.ReplaceAll(in, "_", "/")
	in = strings.ReplaceAll(in, ".", "+")

	output, _ := base64.StdEncoding.DecodeString(in)
	return string(output)
}

// Base64UrlEncode encodes the given input using the URL-safe base64 variant that the WireGuard Portal
// API expects. In comparison to the standard base64 encoding, it uses . instead of +, _ instead of /
// and - instead of =.
func Base64UrlEncode(in string) string {
	out := base64.StdEncoding.EncodeToString([]byte(in))
	out = strings.ReplaceAll(out, "+", ".")
	out = strings.ReplaceAll(out, "/", "_")
	out = strings.ReplaceAll(out, "=", "-")
	return out
}
